cores/arduino/stm32/bootloader.c

```diff
@@ -2,7 +2,144 @@
+  {0x442, 0x1FFFD800},
+  {0x445, 0x1FFFC400},
+  {0x448, 0x1FFFC800},
+  {0x442, 0x1FFFD800},
+#elif STM32F1xx
+  {0x412, 0x1FFFF000},
+  {0x410, 0x1FFFF000},
+  {0x414, 0x1FFFF000},
+  {0x420, 0x1FFFF000},
+  {0x428, 0x1FFFF000},
+  {0x418, 0x1FFFB000},
+  {0x430, 0x1FFFE000},
+#elif STM32F2xx
+  {0x411, 0x1FFF0000},
+#elif STM32F3xx
+  {SYSMEM_ADDR_COMMON, 0x1FFFD800},
+#elif STM32F4xx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#elif STM32F7xx
+  {SYSMEM_ADDR_COMMON, 0x1FF00000},
+#elif STM32G0xx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#elif STM32G4xx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#elif STM32H7xx
+  {0x450, 0x1FF00000},
+#elif STM32L0xx
+  {SYSMEM_ADDR_COMMON, 0x1FF00000},
+#elif STM32L1xx
+  {SYSMEM_ADDR_COMMON, 0x1FF00000},
+#elif STM32L4xx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#elif STM32WBxx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#else
+#warning "No system memory address for this serie!"
+#endif
+  {0x0000, 0x00000000}
+};
+
+uint32_t getSysMemAddr(void)
+{
+  uint32_t sysMemAddr = 0;
+  if (devSysMemAddr[0].devID == SYSMEM_ADDR_COMMON) {
+    sysMemAddr = devSysMemAddr[0].sysMemAddr;
+  } else {
+    uint32_t devId = LL_DBGMCU_GetDeviceID();
+    for (uint32_t id = 0; devSysMemAddr[id].devID != 0; id++) {
+      if (devSysMemAddr[id].devID == devId) {
+        sysMemAddr = devSysMemAddr[id].sysMemAddr;
+        break;
+      }
+    }
+  }
+  return sysMemAddr;
+}
+
+/* Request to jump to system memory boot */
+WEAK void jumpToBootloaderRequested(void)
+{
+  enableBackupDomain();
+  setBackupRegister(SYSBL_MAGIC_NUMBER_BKP_INDEX, SYSBL_MAGIC_NUMBER_BKP_VALUE);
+  NVIC_SystemReset();
+}
+
+/* Jump to system memory boot from user application */
+WEAK void jumpToBootloader(void)
+{
+  enableBackupDomain();
+  if (getBackupRegister(SYSBL_MAGIC_NUMBER_BKP_INDEX) == SYSBL_MAGIC_NUMBER_BKP_VALUE) {
+    setBackupRegister(SYSBL_MAGIC_NUMBER_BKP_INDEX, 0);
+
+#ifdef USBCON
+    USBD_reenumerate();
+#endif
+    void (*sysMemBootJump)(void);
+
+    /**
+     * Get system memory address
+     *
+     * Available in AN2606 document:
+     * Table xxx Bootloader device-dependent parameters
+     */
+    volatile uint32_t sysMem_addr = getSysMemAddr();
+    if (sysMem_addr != 0) {
+#ifdef __HAL_SYSCFG_REMAPMEMORY_SYSTEMFLASH
+      /* Remap system Flash memory at address 0x00000000 */
+      __HAL_SYSCFG_REMAPMEMORY_SYSTEMFLASH();
+#endif
+
+      /**
+       * Set jump memory location for system memory
+       * Use address with 4 bytes offset which specifies jump location
+       * where program starts
+       */
+      sysMemBootJump = (void (*)(void))(*((uint32_t *)(sysMem_addr + 4)));
+
+      /**
+       * Set main stack pointer.
+       * This step must be done last otherwise local variables in this function
+       * don't have proper value since stack pointer is located on different position
+       *
+       * Set direct address location which specifies stack pointer in SRAM location
+       */
+      __set_MSP(*(uint32_t *)sysMem_addr);
```

Comment:
This is already an inline ASM as `__set_MSP()` is defined like this in [cmsis_gcc.h](https://github.com/ARM-software/CMSIS_5/blob/58b1eb3d2d285760af717fc3605813d531b30151/CMSIS/Core/Include/cmsis_gcc.h#L393-L402):
```C
/**
  \brief   Set Main Stack Pointer
  \details Assigns the given value to the Main Stack Pointer (MSP).
  \param [in]    topOfMainStack  Main Stack Pointer value to set
 */
__STATIC_FORCEINLINE void __set_MSP(uint32_t topOfMainStack)
{
  __ASM volatile ("MSR msp, %0" : : "r" (topOfMainStack) : );
}
```

Reply:
Yeah, but I mean that the stack-pointer setting and the jump should be in the same asm block. AFAICS that is the only way to force the compiler to put the jump address in registers before setting the stack pointer (with the current code, the compiler could decide to keep the jump address on the stack and only load it into registers *after* `__set_MSP()` is called).

Reply:
`__set_MSP()` is not called as it is replace inline.

Reply:
Ok, but that does not change my argument. In fact, looking at the disassembly, you can already see that `sysMemAddr` is stored on the stack:

```
 800375a:       f7ff ffb9       bl      80036d0 <getSysMemAddr>
 800375e:       9001            str     r0, [sp, #4]
^^^ This stores the result of getSysMemAddr on the stack
 8003760:       9b01            ldr     r3, [sp, #4]
 8003762:       b183            cbz     r3, 8003786 <jumpToBootloader+0x46>
 8003764:       4b0a            ldr     r3, [pc, #40]   ; (8003790 <jumpToBootloader+0x50>)
 8003766:       681a            ldr     r2, [r3, #0]
 8003768:       f022 0203       bic.w   r2, r2, #3
 800376c:       601a            str     r2, [r3, #0]
 800376e:       681a            ldr     r2, [r3, #0]
 8003770:       f042 0201       orr.w   r2, r2, #1
 8003774:       601a            str     r2, [r3, #0]
 8003776:       9b01            ldr     r3, [sp, #4]
^^^ This loads the value from the stack again
 8003778:       9a01            ldr     r2, [sp, #4]
 800377a:       6812            ldr     r2, [r2, #0]
 800377c:       f382 8808       msr     MSP, r2
 8003780:       685b            ldr     r3, [r3, #4]
^^^ And this does the `*(sysMemAddr + 4)`
 8003782:       4798            blx     r3
```

As you can see, the value is stored on the stack and then loaded just before MSP is updated. It seems rather arbitrary of the compiler to use this order, AFAICS it could just as well have chosen to do set MSP first and then load the value from the stack.

This seems easy to fix, though, just have a bit of inline ASM that handles both the MSP-setting and the jump (though I'm not familiar enough with ARM ASM to just write an example off-hand, I might have a closer look when I have more time later).

Reply:
I found a bit of inline ASM for this that we can probably just copy in the @GrumpyOldPizza STM32L4 core: https://github.com/GrumpyOldPizza/arduino-STM32L4/blob/bacc184288a4644b2ee6a97672334983f3e788ab/system/STM32L4xx/Source/boot_stm32l4xx.c#L159-L166

That core also does quite a bit of setup before jumping to the bootloader (in particular, it seems to lock I/O pins), I'm not sure why all of that is required (since when booting the bootloader using the BOOT pin, none of this happens either). @GrumpyOldPizza, care to comment? :-)